src/write/xcoff.rs

```diff
@@ -0,0 +1,489 @@
+use core::mem;
+
+use crate::endian::{BigEndian as BE, I16, U16, U32};
+use crate::write::string::*;
+use crate::write::util::*;
+use crate::write::*;
+
+use crate::{xcoff, AddressSize};
+
+#[derive(Default, Clone, Copy)]
+struct SectionOffsets {
+    address: usize,
+    data_offset: usize,
+    reloc_offset: usize,
+}
+
+#[derive(Default, Clone, Copy)]
+struct SymbolOffsets {
+    index: usize,
+    str_id: Option<StringId>,
+    aux_count: u8,
+}
+
+impl<'a> Object<'a> {
+    pub(crate) fn xcoff_section_info(
+        &self,
+        section: StandardSection,
+    ) -> (&'static [u8], &'static [u8], SectionKind) {
+        match section {
+            StandardSection::Text => (&[], &b".text"[..], SectionKind::Text),
+            StandardSection::Data => (&[], &b".data"[..], SectionKind::Data),
+            StandardSection::ReadOnlyData
+            | StandardSection::ReadOnlyDataWithRel
+            | StandardSection::ReadOnlyString => (&[], &b".rdata"[..], SectionKind::ReadOnlyData),
+            StandardSection::UninitializedData => {
+                (&[], &b".bss"[..], SectionKind::UninitializedData)
+            }
+            StandardSection::Tls => (&[], &b".tdata"[..], SectionKind::Tls),
+            StandardSection::UninitializedTls => {
+                (&[], &b".tbss"[..], SectionKind::UninitializedTls)
+            }
+            StandardSection::TlsVariables => {
+                // Unsupported section.
+                (&[], &[], SectionKind::TlsVariables)
+            }
+            StandardSection::Common => {
+                // Unsupported section.
+                (&[], &[], SectionKind::Common)
+            }
+        }
+    }
+
+    pub(crate) fn xcoff_fixup_relocation(&mut self, mut relocation: &mut Relocation) -> i64 {
+        let constant = match relocation.kind {
+            RelocationKind::Relative => relocation.addend + 4,
+            _ => relocation.addend,
+        };
+        relocation.addend -= constant;
+        constant
+    }
+
+    pub(crate) fn xcoff_write(&self, buffer: &mut dyn WritableBuffer) -> Result<()> {
```

Comment:
Can we separate writing into 2 phases like ELF does? The 1st phase is to layout where we compute offsets of every items inside the object file; the 2nd phase is to write items in the right places inside the object file. We can also further make the 2nd phase into several subroutines: writing header, writing section headers and etc.

Reply:
It already is two phases. Everything down to line 160 is only computing sizes and offsets.

There's no need to split into separate subroutines like ELF unless we want to expose a lower level XCOFF writer similar to `write::elf::Writer`, but I think that can be done in a later PR if it is needed. For now, keeping it all in one function makes it easier to rely on the compiler to ensure that everything is defined before it is used. This is manageable because the control flow in the function is quite linear.